kratos/utilities/data_type_traits.h

```diff
@@ -0,0 +1,642 @@
+            << rShape << " ].\n";
+        return false;
+    }
+
+    inline static PrimitiveType const * GetContiguousData(const ContainerType& rValue)
+    {
+        return &rValue;
+    }
+
+    inline static PrimitiveType * GetContiguousData(ContainerType& rValue)
+    {
+        return &rValue;
+    }
+
+    inline static void FillToContiguousData(
+        PrimitiveType* pContiguousDataBegin,
+        const ContainerType& rValue)
+    {
+        *pContiguousDataBegin = rValue;
+    }
+
+    inline static void FillFromContiguousData(
+        ContainerType& rValue,
+        PrimitiveType const * pContiguousDataBegin)
+    {
+        rValue = *pContiguousDataBegin;
+    }
+
+    ///@}
+};
+
+template<class TDataType, std::size_t Dimension> class DataTypeTraits<array_1d<TDataType, Dimension>>
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ContainerType = array_1d<TDataType, Dimension>;
+
+    using ValueType = TDataType;
+
+    using PrimitiveType = typename DataTypeTraits<ValueType>::PrimitiveType;
+
+    static constexpr bool HasContiguousPrimitiveData = std::is_same_v<PrimitiveType, ValueType>;
+
+    static constexpr bool HasDynamicMemoryAllocation = DataTypeTraits<ValueType>::HasDynamicMemoryAllocation;
+
+    ///@}
+    ///@name Public static operations
+    ///@{
+
+    static inline unsigned int Size(const ContainerType& rContainer)
+    {
+        return Dimension * DataTypeTraits<ValueType>::Size(rContainer[0]);
+    }
+
+    static inline std::vector<unsigned int> Shape(const ContainerType& rContainer)
+    {
+        auto shape = DataTypeTraits<ValueType>::Shape(rContainer[0]);
+        shape.insert(shape.begin(), Dimension);
+        return shape;
+    }
+
+    static inline bool Reshape(
+        ContainerType& rContainer,
+        const std::vector<unsigned int>& rShape)
+    {
+        KRATOS_ERROR_IF_NOT(rShape.size() >= 1 && rShape[0] == static_cast<unsigned int>(Dimension))
+            << "Invalid shape given for array_1d data type [ Expected shape = [" << Dimension << "], provided shape = "
+            << rShape << " ].\n";
+
+        bool is_reshaped = false;
+
+        if constexpr(DataTypeTraits<ValueType>::HasDynamicMemoryAllocation) {
+            std::vector<unsigned int> sub_data_type_shape(rShape.size() - 1);
+            std::copy(rShape.begin() + 1, rShape.end(), sub_data_type_shape.begin());
+            std::for_each(rContainer.begin(), rContainer.end(), [&is_reshaped, &sub_data_type_shape](auto& rValue) { is_reshaped = DataTypeTraits<ValueType>::Reshape(rValue, sub_data_type_shape) || is_reshaped; });
+        }
+
+        return is_reshaped;
+    }
+
+    inline static PrimitiveType const * GetContiguousData(const ContainerType& rValue)
+    {
+        if constexpr(HasContiguousPrimitiveData) {
+            return rValue.data().begin();
+        } else {
+            static_assert(!std::is_same_v<TDataType, TDataType>, "This should be only called if the rValue is contiguous only.");
```

Comment:
This works as it suppose to be. It is tested, and I have been using it in few other places as well. `static_assert(true)` does not work, it will always throw compiler error irrespective of the compiler branch, but this works.